## Семинар 3. Изменение таблиц в Pandas

## Условие 1: Задача 1
- Скачать данные по ссылке https://www.kaggle.com/datasets/ionaskel/laptop-prices
- Считать данные с помощью pandas
- Вывести на экран первые 5 строк

In [2]:
import pandas as pd

In [12]:
df = pd.read_csv('/kc_house_data.csv', parse_dates=['date'], dayfirst=False)
df.head()

id       date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520 2014-10-13  221900.0         3       1.00         1180   
1  6414100192 2014-12-09  538000.0         3       2.25         2570   
2  5631500400 2015-02-25  180000.0         2       1.00          770   
3  2487200875 2014-12-09  604000.0         4       3.00         1960   
4  1954400510 2015-02-18  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### 1.1 Создать новый признак price_per_sq_lot, который будет содержать среднюю стоимость за один кв.м общей площади

In [13]:
df['price_per_sq_lot'] = (df['price'] / df['sqft_lot']).round(2)
df.head()

id       date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520 2014-10-13  221900.0         3       1.00         1180   
1  6414100192 2014-12-09  538000.0         3       2.25         2570   
2  5631500400 2015-02-25  180000.0         2       1.00          770   
3  2487200875 2014-12-09  604000.0         4       3.00         1960   
4  1954400510 2015-02-18  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  price_per_sq_lot  
0        5650             39.27  
1        7639             74.29  
2        8062             18.00  
3        5000            120.80  
4        7503             63.12  

[5 rows x 22 columns]

### 1.2 Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновацией дома и годом постройки дома
Если реновации дома не было, то в новом признаке поставьте 0


In [14]:
df['delta_renovated'] = 0
df.loc[df['yr_renovated'] != 0, 'delta_renovated'] = df['yr_renovated'] - df['yr_built']
df.sample(16)

id       date      price  bedrooms  bathrooms  sqft_living  \
5254   1898200080 2015-03-12   349000.0         3       2.50         2550   
284    5634500392 2015-04-10   330000.0         3       3.00         2420   
6328   2133020020 2015-01-16   372000.0         3       2.50         1920   
18744   582000065 2014-11-25   725000.0         4       1.75         2700   
19300  2926049564 2014-09-24   360000.0         3       2.25         1381   
16092  8122600020 2014-05-21   200000.0         4       1.00         1310   
9333   3324079089 2014-11-21  1335000.0         4       4.00         5050   
17181  3818700185 2014-09-25   400000.0         4       1.50         2150   
12978   223039229 2014-09-02   787500.0         2       2.50         2390   
1983    434000030 2014-12-19   555000.0         3       2.00         2080   
16510  4060000170 2014-08-20   255000.0         2       1.00         1260   
15962  6699940250 2014-07-25   350000.0         4       2.50         2610   
8333   1525059020 2015-04-10   925000.0         4       2.50         2910   
4231   7443000985 2014-08-15   475000.0         5       2.50         2010   
4325   3876300600 2014-12-12   371000.0         4       1.75         1610   
21562  2122059216 2015-04-14   422000.0         4       2.50         2930   

       sqft_lot  floors  waterfront  view  ...  sqft_basement  yr_built  \
5254       7709     2.0           0     0  ...              0      1989   
284       13959     1.0           0     0  ...            680      1988   
6328      15260     2.0           0     0  ...              0      1990   
18744      6000     1.0           0     0  ...           1250      1953   
19300      1180     3.0           0     0  ...              0      2007   
16092      5200     1.5           0     0  ...            150      1945   
9333     202554     2.0           0     0  ...           1790      2000   
17181     11026     1.0           0     0  ...              0      1952   
12978      6928     2.0           0     3  ...            580      1949   
1983       7020     1.0           0     0  ...           1040      1951   
16510      7810     2.0           0     0  ...              0      1945   
15962      5866     2.0           0     0  ...              0      2005   
8333      48351     1.0           0     0  ...           1000      1967   
4231       3600     1.5           0     0  ...            500      1912   
4325      11305     1.0           0     0  ...              0      1968   
21562      5973     2.0           0     0  ...              0      2008   

       yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  \
5254              0    98023  47.3081 -122.391           2410        9250   
284               0    98028  47.7486 -122.230           2570       13300   
6328              0    98019  47.7317 -121.964           2370       15235   
18744             0    98199  47.6539 -122.395           2080        6000   
19300             0    98125  47.7110 -122.320           1381        1180   
16092             0    98126  47.5384 -122.370           1090        5180   
9333              0    98027  47.5269 -121.922           3370      213444   
17181             0    98028  47.7635 -122.263           1760       10283   
12978          1982    98146  47.5101 -122.391           2390       12852   
1983              0    98115  47.6768 -122.285           1920        7000   
16510             0    98178  47.5003 -122.248           1260        7755   
15962             0    98038  47.3441 -122.040           2480        5188   
8333              0    98005  47.6495 -122.164           2760       43560   
4231              0    98119  47.6522 -122.366           1780        3600   
4325              0    98034  47.7270 -122.176           1870       11850   
21562             0    98030  47.3846 -122.186           3038        7095   

       price_per_sq_lot  delta_renovated  
5254              45.27                0  
284               23.64          

In [8]:
delta = []
for i, row in df.iterrows():
  if row['yr_renovated']:
      delta.append(row['yr_renovated'] - row['yr_built'])
  else:
      delta.append(0)
df['delta_renovated'] = delta
df.sample(16)

id       date     price  bedrooms  bathrooms  sqft_living  \
11453  7972604001 2014-07-18  354000.0         5       1.75         1830   
3066   9358001566 2015-02-13  400000.0         2       2.50         1410   
13865  8651410120 2015-03-18  225500.0         3       1.00         1100   
10867    91000135 2015-05-07  750000.0         4       1.50         2060   
4381   8109800110 2014-08-01  717550.0         3       3.50         2840   
17650  3904930730 2014-11-05  496600.0         3       2.50         1910   
5855   2877103070 2015-02-23  775000.0         4       2.50         2040   
20466  6639900242 2014-10-03  750000.0         4       2.50         2850   
21483  2776600082 2014-11-13  407500.0         3       3.50         1522   
18619  3856900590 2014-08-06  640000.0         4       1.75         2100   
1957   8849300160 2015-02-05  345000.0         4       2.50         2280   
18627  9264921020 2014-09-08  260000.0         3       1.50         1750   
16161  3034200426 2015-03-20  450500.0         3       1.00         1410   
5705   2624079022 2014-10-20  530000.0         3       2.25         1880   
18402  6329000385 2014-06-18  825000.0         4       3.50         3810   
7208   1446403145 2015-01-22  170000.0         2       1.00         1240   

       sqft_lot  floors  waterfront  view  ...  sqft_basement  yr_built  \
11453      7986     1.0           0     0  ...            770      1962   
3066       1281     2.0           0     0  ...            320      2008   
13865      5200     1.0           0     0  ...              0      1969   
10867      4000     1.5           0     2  ...            480      1920   
4381       4468     3.0           0     0  ...              0      2006   
17650      5562     2.0           0     0  ...              0      1988   
5855       5000     1.5           0     0  ...            860      1924   
20466     12429     2.0           0     0  ...              0      2008   
21483      1465     2.0           0     0  ...            274      2006   
18619      3000     1.5           0     0  ...            600      1911   
1957       8190     1.0           0     3  ...            890      1983   
18627      7000     2.0           0     0  ...              0      1983   
16161      9384     1.0           0     0  ...              0      1948   
5705     100623     1.5           0     0  ...              0      1987   
18402      9792     2.0           0     0  ...              0      1938   
7208       9900     1.0           0     0  ...            300      1950   

       yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  \
11453             0    98106  47.5208 -122.350           1410        7260   
3066              0    98126  47.5659 -122.370           1410        2550   
13865             0    98042  47.3643 -122.081            920        5200   
10867          1990    98103  47.6857 -122.353           1160        4000   
4381              0    98052  47.7069 -122.117           3040        5400   
17650             0    98029  47.5738 -122.017           1940        4647   
5855              0    98117  47.6786 -122.360           1540        5000   
20466             0    98033  47.6915 -122.177           2540       12000   
21483             0    98117  47.6922 -122.375           1522        1341   
18619             0    98103  47.6721 -122.329           1690        4000   
1957              0    98188  47.4414 -122.272           1990        9000   
18627             0    98023  47.3108 -122.346           1840        9305   
16161             0    98133  47.7161 -122.330           1990        9384   
5705           2007    98024  47.5342 -121.883           2520       21689   
18402          2013    98146  47.5018 -122.380           1950        9792   
7208              0    98168  47.4862 -122.326           1050        9375   

       price_per_sq_lot  delta_renovated  
11453         44.327573                0  
3066         312.256050                0  
13865  

### 1.3 Создайте признаки года продажи, месяца продажи


In [ ]:
df['yr_sold'] = df['date'].dt.year
df['month_sold'] = df['date'].dt.month_name()
df.sample(10)

id       date      price  bedrooms  bathrooms  sqft_living  \
13729  1423049029 2015-03-06   265000.0         4       1.75         1970   
5306   8087800020 2014-05-15   412500.0         3       1.50         1490   
4120    952004875 2014-06-02   661000.0         4       2.25         1990   
17305   452002135 2015-04-22  1070000.0         4       2.50         2740   
1823   4322500055 2015-04-17   607000.0         5       1.75         1910   
11111  8856920550 2015-03-19   378000.0         3       2.50         2130   
1297   2492200435 2014-06-06   389250.0         2       1.50         1490   
636    6131600060 2014-08-15   214000.0         3       1.00         1200   
10857  2310030390 2015-04-29   275000.0         3       1.75         1180   
11911  9828700685 2015-05-13   900000.0         3       1.75         1540   

       sqft_lot  floors  waterfront  view  ...  yr_built  yr_renovated  \
13729      8390     1.0           0     0  ...      1955             0   
5306       8475     1.0           0     0  ...      1963             0   
4120       4600     1.5           0     2  ...      1932             0   
17305      5000     2.0           0     0  ...      2012             0   
1823       5428     1.0           0     0  ...      1954             0   
11111      8404     2.0           0     0  ...      1991             0   
1297       4080     1.0           0     0  ...      1956             0   
636        8316     1.0           0     0  ...      1953             0   
10857      6260     1.0           0     0  ...      1993             0   
11911      8400     1.5           0     0  ...      1902             0   

       zipcode      lat     long  sqft_living15  sqft_lot15  delta_renovated  \
13729    98178  47.4861 -122.251           1710       10890                0   
5306     98052  47.6571 -122.133           1490        8540                0   
4120     98126  47.5638 -122.380           1810        5750                0   
17305    98107  47.6740 -122.371           1660        5000                0   
1823     98136  47.5333 -122.392           1820        5900                0   
11111    98058  47.4623 -122.130           2130        8404                0   
1297     98126  47.5344 -122.381           1320        4080                0   
636      98002  47.3221 -122.215           1200        8316                0   
10857    98038  47.3531 -122.048           1580        7272                0   
11911    98122  47.6187 -122.295           1330        4800                0   

       yr_sold  month_sold  
13729     2015       March  
5306      2014         May  
4120      2014        June  
17305     2015       April  
1823      2015       April  
11111     2015       March  
1297      2014        June  
636       2014      August  
10857     2015       April  
11911     2015         May  

[10 rows x 24 columns]

### 1.4 Удалите признаки date, zipcode, lat, long

In [ ]:
df.drop(columns=['date', 'zipcode', 'lat', 'long'], inplace=True)
df.sample(10)

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
8987   5035300650  1099500.0         4       2.00         2580      6000   
16604  1432700420   250000.0         3       1.00         1460     10914   
4174   2461900945   438000.0         5       1.00         1950      6250   
7418   3881900321   339950.0         3       1.00         1050      5402   
13722  4039700870  1160000.0         5       1.75         2870      9680   
13674  6163900333   338000.0         3       1.75         1250      7710   
2556   2973800030   175000.0         3       1.00         1030      6600   
2039   3395040550   320000.0         3       2.50         1530      2890   
14877  3274850130   392000.0         4       2.50         2600      8921   
3145   9274201807   580000.0         3       2.50         1590      1937   

       floors  waterfront  view  condition  grade  sqft_above  sqft_basement  \
8987      1.0           0     0          5      9        1300           1280   
16604     1.0           0     0          4      7        1460              0   
4174      1.5           0     0          4      7        1450            500   
7418      1.5           0     0          4      7        1050              0   
13722     1.0           0     4          5      9        1440           1430   
13674     1.0           0     0          4      7        1250              0   
2556      1.0           0     0          3      7        1030              0   
2039      2.0           0     0          3      7        1530              0   
14877     2.0           0     0          3      8        2600              0   
3145      2.5           0     0          3      8        1590              0   

       yr_built  yr_renovated  sqft_living15  sqft_lot15  delta_renovated  \
8987       1950             0           2300        6200                0   
16604      1959             0           1490        8314                0   
4174       1917             0           1950        6250                0   
7418       1906             0           1510        4685                0   
13722      1966             0           2940        9729                0   
13674      1947             0           1340        7710                0   
2556       1954             0           1220        9040                0   
2039       2001             0           1540        2890                0   
14877      1991             0           2550        8683                0   
3145       2004             0           1620        2022                0   

       yr_sold month_sold  
8987      2014       June  
16604     2014       July  
4174      2015      April  
7418      2014        May  
13722     2015   February  
13674     2014   November  
2556      2014       July  
2039      2015      April  
14877     2014   December  
3145      2014     August

## Условие 2: 2 задача
Создайте датафрейм с клиентами:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})


In [ ]:
clients = pd.DataFrame({
    'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295, 939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242, 2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
    'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060, 1189000207, 6821600300, 7137950720, 9510920050, 6131600255, 5428000070, 1788800910, 8100400160, 3123049142, 6306800010, 5083000375,
                 920100025, 1951600150, 809001400, 339600110, 1622049154, 1099600250, 8563000110, 2768100205, 3995700435, 8861700030, 3303980210, 7731100066, 8146100580, 825069097, 3889100029, 9524100196]
})
clients.head()

client_id    house_id
0       1459  8965450190
1       4684  6823100225
2       3498  5104540330
3       3942  2131701075
4       4535  1522700060

## 2.1 Присоедините к таблице clients данные по домам через метод join


In [ ]:
cl_id = clients.set_index('house_id')
cl_id.head()

client_id
house_id             
8965450190       1459
6823100225       4684
5104540330       3498
2131701075       3942
1522700060       4535

In [ ]:
df_id = df.set_index('id')
df.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0     0          3      7        1180              0      1955   
1           0     0          3      7        2170            400      1951   
2           0     0          3      6         770              0      1933   
3           0     0          5      7        1050            910      1965   
4           0     0          3      8        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  delta_renovated  yr_sold  \
0             0           1340        5650                0     2014   
1          1991           1690        7639               40     2014   
2             0           2720        8062                0     2015   
3             0           1360        5000                0     2014   
4             0           1800        7503                0     2015   

  month_sold  
0    October  
1   December  
2   February  
3   December  
4   February

In [ ]:
cl_id.join(df_id).sample(10)

client_id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
3123049142        280  452000.0       3.0       2.25       2600.0   14810.0   
6823100225       4684  700000.0       4.0       1.75       1870.0    6000.0   
8965450190       1459  295000.0       3.0       2.50       1500.0    3060.0   
8563000110       2068  427000.0       4.0       1.75       1460.0    9750.0   
5083000375       4332  170000.0       3.0       1.00       1310.0    9529.0   
809001400        3113  925000.0       3.0       1.00       1630.0    3200.0   
9524100196       1474  239000.0       2.0       1.50        680.0     772.0   
1522700060       4535  518000.0       4.0       2.75       2520.0   14021.0   
920100025        3475       NaN       NaN        NaN          NaN       NaN   
5428000070       2295  770000.0       5.0       3.50       4750.0    8234.0   

            floors  waterfront  view  condition  grade  sqft_above  \
3123049142     1.0         0.0   2.0        4.0    8.0      1490.0   
6823100225     1.0         0.0   0.0        5.0    8.0      1670.0   
8965450190     2.0         0.0   0.0        3.0    7.0      1500.0   
8563000110     1.0         0.0   0.0        4.0    7.0      1460.0   
5083000375     1.0         0.0   0.0        3.0    7.0      1310.0   
809001400      1.5         0.0   0.0        4.0    8.0      1630.0   
9524100196     2.0         0.0   0.0        3.0    7.0       680.0   
1522700060     2.0         0.0   0.0        3.0    9.0      2520.0   
920100025      NaN         NaN   NaN        NaN    NaN         NaN   
5428000070     2.0         0.0   2.0        3.0   10.0      3350.0   

            sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
3123049142         1110.0    1956.0           0.0         2450.0     16715.0   
6823100225          200.0    1949.0           0.0         1710.0      6000.0   
8965450190            0.0    1994.0           0.0         2700.0      7734.0   
8563000110            0.0    1967.0           0.0         1820.0      9840.0   
5083000375            0.0    1956.0           0.0         1330.0      9529.0   
809001400             0.0    1912.0           0.0         1710.0      3600.0   
9524100196            0.0    2005.0           0.0          690.0      1059.0   
1522700060            0.0    1999.0           0.0         2330.0     14007.0   
920100025             NaN       NaN           NaN            NaN         NaN   
5428000070         1400.0    2013.0           0.0         2160.0     14496.0   

            delta_renovated  yr_sold month_sold  
3123049142              0.0   2014.0     August  
6823100225              0.0   2015.0      April  
8965450190              0.0   2015.0   February  
8563000110              0.0   2015.0      April  
5083000375              0.0   2014.0    October  
809001400               0.0   2014.0  September  
9524100196              0.0   2014.0   November  
1522700060              0.0   2014.0       June  
920100025               NaN      NaN        NaN  
5428000070              0.0   2015.0        May

## 2.2 Присоедините к таблице clients данные по домам через метод merge. Это нужно, чтобы понимать, какие дома покупались клиентами. house_id - это индексы датафрейма с домами

In [ ]:
clients.merge(df, left_on='house_id', right_on='id').head()

client_id    house_id          id     price  bedrooms  bathrooms  \
0       1459  8965450190  8965450190  295000.0         3       2.50   
1       4684  6823100225  6823100225  700000.0         4       1.75   
2       3498  5104540330  5104540330  679000.0         4       2.50   
3       3942  2131701075  2131701075  420000.0         3       1.75   
4       4535  1522700060  1522700060  518000.0         4       2.75   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1500      3060     2.0           0  ...      7        1500   
1         1870      6000     1.0           0  ...      8        1670   
2         3680      7236     2.0           0  ...     10        3680   
3         1720      5000     1.5           0  ...      8        1720   
4         2520     14021     2.0           0  ...      9        2520   

   sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
0              0      1994             0           2700        7734   
1            200      1949             0           1710        6000   
2              0      2006             0           3310        7180   
3              0      1932          2009           1410        8300   
4              0      1999             0           2330       14007   

   delta_renovated  yr_sold  month_sold  
0                0     2015    February  
1                0     2015       April  
2                0     2015         May  
3               77     2014    December  
4                0     2014        June  

[5 rows x 22 columns]

## Условие 3: 3 задача - Составьте несколько сводных таблиц
### 3.1 Найдите среднюю стоимость домов в зависимости от количества спален. Отсортируйте от меньшей стоимости к большей


In [ ]:
df.groupby('bedrooms').agg({'price': 'mean'}).sort_values('price').round(2)

price
bedrooms            
1          317642.88
2          401372.68
0          409503.85
3          466232.08
11         520000.00
4          635419.50
33         640000.00
5          786599.83
10         819333.33
6          825520.64
9          893999.83
7          951184.66
8         1105076.92

### 3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома


In [ ]:
df.groupby('condition').agg({'price': ['min', 'mean', 'max']}).round(2)

price                      
                min       mean        max
condition                                
1           78000.0  334431.67  1500000.0
2           80000.0  327287.15  2555000.0
3           75000.0  542012.58  7062500.0
4           89000.0  521200.39  7700000.0
5          110000.0  612418.09  3650000.0

### 3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида


In [ ]:
df.pivot_table(index='waterfront', columns='view', values='id', aggfunc='count', fill_value=0)

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

In [ ]:
df.groupby(['view', 'waterfront']).agg({'id': 'count'}).unstack('view').fillna(0)

id                            
view              0      1      2      3      4
waterfront                                     
0           19489.0  331.0  955.0  491.0  184.0
1               0.0    1.0    8.0   19.0  135.0

In [15]:
pd.crosstab(index=df['waterfront'], columns=df['view'])

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

### 3.4 Каких домов в зависимости от этажности и количества спален больше?


In [ ]:
pd.crosstab(index=df['floors'], columns=df['bedrooms'])

bedrooms  0    1     2     3     4    5    6   7   8   9   10  11  33
floors                                                               
1.0        4  162  1951  5455  2383  605  104   9   5   0   1   0   1
1.5        0   21   182   786   698  185   30   7   1   0   0   0   0
2.0        6   12   497  3118  3682  775  119  19   6   4   2   1   0
2.5        0    1     5    56    58   23   14   2   0   2   0   0   0
3.0        2    3   123   405    61   13    5   1   0   0   0   0   0
3.5        1    0     2     4     0    0    0   0   1   0   0   0   0

Домов с 1 этажом и 3 спальнями больше (5455).

### 3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома

In [ ]:
df.pivot_table(index='condition', columns='grade', values='price', aggfunc='median', fill_value=0)

grade          1       3       4       5       6       7       8        9   \
condition                                                                    
1          142000       0  150000  190000  255000  403500  932500        0   
2               0  280000  145000  180000  235000  305000  429000   715000   
3               0   75000  205000  234475  265000  357500  485000   689000   
4               0       0  238525  229975  295000  390000  571250   823500   
5               0  262000       0  227450  285475  456000  696000  1078000   

grade           10       11       12       13  
condition                                      
1                0        0        0        0  
2          1752500        0        0        0  
3           890000  1209500  1807500  2888000  
4          1030000  1685000  2125000  5750000  
5          1650000  2050000  1990000        0

In [ ]:
df.groupby(['condition', 'grade']).agg({'price': 'median'}).unstack('grade').fillna(0)

price                                                    \
grade            1         3         4         5         6         7    
condition                                                               
1          142000.0       0.0  150000.0  190000.0  255000.0  403500.0   
2               0.0  280000.0  145000.0  180000.0  235000.0  305000.0   
3               0.0   75000.0  205000.0  234475.0  265000.0  357500.0   
4               0.0       0.0  238525.0  229975.0  295000.0  390000.0   
5               0.0  262000.0       0.0  227450.0  285475.0  456000.0   

                                                                            
grade            8          9          10         11         12         13  
condition                                                                   
1          932500.0        0.0        0.0        0.0        0.0        0.0  
2          429000.0   715000.0  1752500.0        0.0        0.0        0.0  
3          485000.0   689000.0   890000.0  1209500.0  1807500.0  2888000.0  
4          571250.0   823500.0  1030000.0  1685000.0  2125000.0  5750000.0  
5          696000.0  1078000.0  1650000.0  2050000.0  1990000.0        0.0